cmake/CMakeLists.txt

```diff
@@ -998,7 +1010,7 @@ function(onnxruntime_set_compile_flags target_name)
       target_compile_options(${target_name} PRIVATE "$<$<COMPILE_LANGUAGE:CUDA>:SHELL:--compiler-options -Wno-deprecated-copy>" "$<$<COMPILE_LANGUAGE:CXX>:-Wno-deprecated-copy>")
     endif()
     if(onnxruntime_USE_CUDA)
-      if(HAS_STRICT_ALIASING)
+      if((NVCC_HAS_STRICT_ALIASING AND "${target_name}" MATCHES "cuda")  OR (HAS_STRIC_ALIASING AND NOT "${target_name}" MATCHES "cuda"))
```

Comment:
The following line only applies to *.cu files, so I think this "if" statement is not needed.
BTW, if you have fixed the warning, do we still need to disable it? Probably you can totally remove line 1013 to line 1015.

Reply:
Actually we do need this check. How can we just add a flag to the list of options? onnxruntime_USE_CUDA is set by the command line by specifying "--use_cuda". That does not mean all the code we compile target GPU.
What are we doing here? We are deciding whether to add "-Wno-strict-aliasing" my option to the list of compiler options or not. For that we first need to check if the compiler we are using (gxx, nvcc, or something else) supports the flag or not. We are setting HAS_STRIC_ALIASING/NVCC_HAS_STRICT_ALIASING flags using check_cxx_compiler_flag/check_nvcc_compiler_flag. Also based on the target we are building we infer that compiler name. In this case if the target_name matches "cuda" then we "know" that we are using the NVCC compiler and gxx otherwise.
If the NVCC we are using for the compilation does support this flag then only we should add the flag. That is the problem we are trying to solve. This is not just warning, the build fails with the fatal error when nvcc doesn't support the specified flag, in this case '-Wno-strict-aliasing'.

Reply:
I don't know if we can do without this flag. That is something to consider. We might run into other issue with GXX if we don't specify this flag. I made the necessary changes to address compilation error.

Reply:
The following is the error on the master.
nvcc fatal   : Unknown option '-Wno-strict-aliasing'
CMakeFiles/onnxruntime_test_cuda_ops_lib.dir/build.make:62: recipe for target 'CMakeFiles/onnxruntime_test_cuda_ops_lib.dir/bert_ort/sajandhy/onnxruntime/onnxruntime/test/shared_lib/cuda_ops.cu.o' failed